websocket/jetty12-ee10/src/main/java/jenkins/websocket/Jetty12EE10Provider.java

```diff
@@ -0,0 +1,172 @@
+ * THE SOFTWARE.
+ */
+
+package jenkins.websocket;
+
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.Future;
+import org.eclipse.jetty.ee10.websocket.server.JettyServerUpgradeRequest;
+import org.eclipse.jetty.ee10.websocket.server.JettyServerUpgradeResponse;
+import org.eclipse.jetty.ee10.websocket.server.JettyWebSocketServerContainer;
+import org.eclipse.jetty.websocket.api.Callback;
+import org.eclipse.jetty.websocket.api.Session;
+import org.kohsuke.MetaInfServices;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+@Restricted(NoExternalUse.class)
+@MetaInfServices(Provider.class)
+public class Jetty12EE10Provider implements Provider {
+
+    /**
+     * Number of seconds a WebsocketConnection may stay idle until it expires. Zero to disable. This
+     * value must be higher than the <code>jenkins.websocket.pingInterval</code>. Per <a
+     * href=https://eclipse.dev/jetty/documentation/jetty-12/programming-guide/index.html#pg-websocket-session-ping>Jetty
+     * 12 documentation</a> a ping mechanism should keep the websocket active. Therefore, the idle
+     * timeout must be higher than the ping interval to avoid timeout issues.
+     */
+    private static long IDLE_TIMEOUT_SECONDS = Long.getLong("jenkins.websocket.idleTimeout", 60L);
+
+    private static final String ATTR_LISTENER = Jetty12EE10Provider.class.getName() + ".listener";
+
+    private boolean initialized = false;
+
+    public Jetty12EE10Provider() {
+        JettyWebSocketServerContainer.class.hashCode();
+    }
+
+    private void init(HttpServletRequest req) {
+        if (!initialized) {
+            JettyWebSocketServerContainer.getContainer(req.getServletContext()).setIdleTimeout(Duration.ofSeconds(IDLE_TIMEOUT_SECONDS));
+            initialized = true;
+        }
+    }
+
+    @Override
+    public Handler handle(HttpServletRequest req, HttpServletResponse rsp, Listener listener) throws Exception {
+        init(req);
+        req.setAttribute(ATTR_LISTENER, listener);
+        // TODO Jetty 10+ has no obvious equivalent to WebSocketServerFactory.isUpgradeRequest; RFC6455Negotiation?
+        if (!"websocket".equalsIgnoreCase(req.getHeader("Upgrade"))) {
+            rsp.sendError(HttpServletResponse.SC_BAD_REQUEST, "only WS connections accepted here");
+            return null;
+        }
+        if (!JettyWebSocketServerContainer.getContainer(req.getServletContext()).upgrade(Jetty12EE10Provider::createWebSocket, req, rsp)) {
+            rsp.sendError(HttpServletResponse.SC_BAD_REQUEST, "did not manage to upgrade");
+            return null;
+        }
+        return new Handler() {
+            @Override
+            public Future<Void> sendBinary(ByteBuffer data) {
+                Callback.Completable f = new Callback.Completable();
```

Comment:
From [the migration guide](https://jetty.org/docs/jetty/12/programming-guide/migration/11-to-12.html#websocket):

> 

The Jetty 11 `WriteCallback` class has been renamed to just `Callback` in Jetty 12, because it is now also used when receiving binary data. Note that this `Callback` interface is a different interface from the `org.eclipse.jetty.util.Callback` interface, which cannot be used in the Jetty WebSocket APIs due to class loader visibility issues.